spec/requests/flows/device_code_spec.rb

```diff
@@ -0,0 +1,128 @@
+    before do
+      config_is_set(:device_verification_url, 'http://www.example.com/devices')
+      config_is_set(:device_polling_interval, 5)
+      config_is_set(:grant_flows, ['device_code'])
+    end
+
+    context 'for grant_type device_code' do
+      before do
+        config_is_set(:grant_flows, ['device_code'])
+      end
+
+      it 'returns the device code response' do
+        params = {
+          grant_type: 'device_code',
+          client_id: client.uid
+        }
+
+        post '/oauth/device', params
+
+        expect(response.status).to eq(200)
+
+        access_grant = Doorkeeper::DeviceAccessGrant.first
+        should_have_json 'code', access_grant.token
+        should_have_json 'user_code', access_grant.user_token
+        should_have_json 'verification_url', 'http://www.example.com/devices'
+        should_have_json 'expires_in', access_grant.expires_in
+        should_have_json 'interval', 5
+      end
+    end
+
+    context 'for grant_type device_token' do
+      let(:device_access_grant) { FactoryGirl.create(:device_access_grant, application: client) }
+      let(:params) do
+        {
+          grant_type: 'device_token',
+          client_id: client.uid,
+          client_secret: client.secret,
+          code: device_access_grant.token
+        }
+      end
+
+      before do
+        config_is_set(:grant_flows, ['device_token'])
+      end
+
+      context 'when user has authorized application' do
+        let(:user) { FactoryGirl.create(:user) }
+
+        before do
+          device_access_grant.update_attribute(:resource_owner_id, user.id)
+        end
+
+        it 'returns the access token' do
+          post '/oauth/device', params
+
+          expect(response.status).to eq(200)
+
+          access_token = Doorkeeper::AccessToken.first
+          should_have_json 'access_token', access_token.token
+          should_have_json 'token_type', 'bearer'
+          should_have_json 'expires_in', access_token.expires_in
+        end
+      end
+
+      context 'when user has not yet authorized application' do
+        it 'returns an error' do
+          post '/oauth/device', params
+
+          expect(response.status).to eq(401)
+
+          should_have_json 'error', 'authorization_pending'
+          should_have_json 'error_description', translated_error_message('authorization_pending')
+        end
+      end
+
+      context 'when user declined to authorize application' do
+        before do
+          device_access_grant.destroy
+        end
+
+        it 'returns an error' do
+          post '/oauth/device', params
+
+          expect(response.status).to eq(401)
+
+          should_have_json 'error', 'authorization_declined'
+          should_have_json 'error_description', translated_error_message('authorization_declined')
+        end
+      end
+
+      context 'when access grant is expired' do
+        before do
+          device_access_grant.update_attribute(:revoked_at, Time.now)
```

Comment:
Do not use `Time.now` without zone. Use one of `Time.zone.now`, `Time.now.current`, `Time.now.in_time_zone`, `Time.now.utc`, `Time.now.getlocal`, `Time.now.iso8601`, `Time.now.jisx0301`, `Time.now.rfc3339`, `Time.now.to_i`, `Time.now.to_f` instead.